hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/store/StoreFactory.java

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright (C) 2024 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hedera.node.app.spi.store;
+
+import com.hedera.hapi.node.transaction.TransactionBody;
+import edu.umd.cs.findbugs.annotations.NonNull;
+
+/**
+ * Factory for creating stores and service APIs.
+ */
+public interface StoreFactory {
+
+    /**
+     * Get a readable store given the store's interface. This gives read-only access to the store.
+     *
+     * @param storeInterface The store interface to find and create a store for
+     * @param <T> Interface class for a Store
+     * @return An implementation of the provided store interface
+     * @throws IllegalArgumentException if the storeInterface class provided is unknown to the app
+     * @throws NullPointerException if {@code storeInterface} is {@code null}
+     */
+    @NonNull
+    <T> T readableStore(@NonNull Class<T> storeInterface);
+
+    /**
+     * Return a writable store given the store's interface. This gives write access to the store.
+     *
+     * <p>This method is limited to stores that are part of the transaction's service.
+     *
+     * @param storeInterface The store interface to find and create a store for
+     * @param <T> Interface class for a Store
+     * @return An implementation of the provided store interface
+     * @throws IllegalArgumentException if the storeInterface class provided is unknown to the app
+     * @throws NullPointerException if {@code storeInterface} is {@code null}
+     */
+    @NonNull
+    <T> T writableStore(@NonNull Class<T> storeInterface);
+
+    /**
+     * Return a service API given the API's interface. This permits use of another service
+     * that doesn't have a corresponding HAPI {@link TransactionBody}.
+     *
+     * @param apiInterface The API interface to find and create an implementation of
+     * @param <T> Interface class for an API
+     * @return An implementation of the provided API interface
+     * @throws IllegalArgumentException if the apiInterface class provided is unknown to the app
+     * @throws NullPointerException if {@code apiInterface} is {@code null}
+     */
+    @NonNull
+    <T> T serviceApi(@NonNull Class<T> apiInterface);
```

Comment:
Based on the name `StoreFactory` and the package `store`, I'm surprised to see `serviceApi` in this interface. Maybe this isn't just a `StoreFactory` and maybe it doesn't belong exclusively to the `store`?

Reply:
I agree that the name `StoreFactory` does not capture the intent. I have not found a better name yet. It probably has to mature a bit. 🙂